cpp/serve/grammar/grammar_impl.h

```diff
@@ -0,0 +1,170 @@
+ * - Or rule: a choice of subrules, e.g. ("a" "b") | "c". Each subrule can be matched.
+ *
+ * Every subrule is represented by a variable-length vector of TData (int32_t), where the first
+ * element indicates the type of the subrule, and the subsequent elements represent the data of the
+ * subrule. We store all subrules in one vector and record the starting position of each subrule. At
+ * the same time, each subrule corresponds to an subrule id for reference.
+ *
+ * For the format in every subrule, see BNFGrammarImpl::DataKind.
+ */
+class BNFGrammarImpl : public BNFGrammarNode {
+ public:
+  /*! \brief The id of a rule. Refers to the index of a rule in the rule vector. */
+  using TRuleId = int32_t;
+  /*! \brief The id of a subrule. */
+  using TSubruleId = int32_t;
+
+  /****************** Rule handling ******************/
+  /*! \brief A rule with name. */
+  struct Rule {
+    std::string name;
+    TSubruleId subrule;
+  };
+
+  /*! \brief Insert a rule and return the rule id. */
+  TRuleId InsertRule(const Rule& rule) {
+    rules_.push_back(rule);
+    return rules_.size() - 1;
+  }
+
+  /*! \brief Get the rule with the given id. */
+  Rule& operator[](TRuleId rule_id) { return rules_[rule_id]; }
+
+  /****************** Subrule handling ******************/
+  /*! \brief The data type of the content of subrules. */
+  using TData = int32_t;
+
+  enum class DataKind : TData {
+    // Format: [kCharacterRange, lower0, upper0, lower1, upper1, ...]
+    // to represent a single character, just add the same lower and upper bound.
+    kCharacterRange,
+    // Format: [kNotCharacterRange, lower0, upper0, lower1, upper1, ...]
+    kNotCharacterRange,
+    // Format: [kEmpty]
+    kEmpty,
+    // Format: [kRuleRef, rule_id]
+    kRuleRef,
+    // Format: [kSequence, subrule_id0, subrule_id1, ...]
+    kSequence,
+    // Format: [kOrRule, subrule_id0, subrule_id1, ...]
+    kOrRule,
+  };
+
+  /*! \brief Stores the data of all subrules. */
+  class SubruleStorage {
+   public:
+    TSubruleId InsertSubrule(const std::vector<TData>& data) {
+      start_index_.push_back(data_.size());
+      data_.insert(data_.end(), data.begin(), data.end());
+      return start_index_.size() - 1;
+    }
+
+    std::pair<const TData*, const TData*> GetSubrule(TSubruleId subrule_id) const {
+      if (subrule_id == start_index_.size() - 1) {
+        return {data_.data() + start_index_[subrule_id], data_.data() + data_.size()};
+      }
+      return {data_.data() + start_index_[subrule_id], data_.data() + start_index_[subrule_id + 1]};
+    }
+
+   private:
+    std::vector<TData> data_;
+    std::vector<TSubruleId> start_index_;
+    friend class BNFGrammarImpl;
+    friend class BNFGrammar;
+  };
+
+  /*! \brief One element of a character range, containing a lower and a upper bound.*/
+  struct CharacterRangeElement {
+    int32_t lower;
+    int32_t upper;
+  };
```

Comment:
Both the lower and upper bound are inclusive right? Might be good to note that the upper bound is inclusive or exclusive.

Reply:
Yeah ok